File: ports/go/sqlsec/sqlsec_test.go
package sqlsec

import "testing"

func hasRule(fs []Finding, id string) bool {
	for _, f := range fs {
		if f.RuleID == id {
			return true
		}
	}
	return false
}

func TestConcatFlaggedSQL001(t *testing.T) {
	src := `query := "SELECT * FROM users WHERE id = " + userID`
	fs := ScanText(src)
	if !hasRule(fs, "SQL001") {
		t.Fatalf("expected SQL001, got %+v", fs)
	}
}

func TestSprintfFlaggedSQL002(t *testing.T) {
	src := `q := fmt.Sprintf("SELECT * FROM t WHERE name = '%s'", name)`
	fs := ScanText(src)
	if !hasRule(fs, "SQL002") {
		t.Fatalf("expected SQL002, got %+v", fs)
	}
}

func TestPercentFormatSQL003(t *testing.T) {
	src := `q = "SELECT * FROM t WHERE id = %s" % value`
	fs := ScanText(src)
	if !hasRule(fs, "SQL003") {
		t.Fatalf("expected SQL003, got %+v", fs)
	}
}

func TestNonSQLConcatIgnored(t *testing.T) {
	src := `greeting := "hello " + name`
	if fs := ScanText(src); len(fs) != 0 {
		t.Fatalf("expected no findings for non-SQL concat, got %+v", fs)
	}
}

func TestParameterizedIsClean(t *testing.T) {
	src := `db.Query("SELECT * FROM users WHERE id = ?", userID)`
	if fs := ScanText(src); len(fs) != 0 {
		t.Fatalf("expected clean parameterized query, got %+v", fs)
	}
}

func TestLineNumbersReported(t *testing.T) {
	src := "package main\n\nq := \"SELECT * FROM t WHERE x=\" + v"
	fs := ScanText(src)
	if len(fs) == 0 || fs[0].Line != 3 {
		t.Fatalf("expected finding on line 3, got %+v", fs)
	}
}

func TestEmptyInput(t *testing.T) {
	if fs := ScanText(""); len(fs) != 0 {
		t.Fatalf("expected no findings on empty input, got %+v", fs)
	}
}

File: ports/ts/test/sqlsec.test.ts
import { strict as assert } from "node:assert";
import test from "node:test";
import { scanText, Finding } from "../src/sqlsec";

const has = (fs: Finding[], id: string) => fs.some((f) => f.ruleId === id);

test("concatenation flagged as SQL001", () => {
  const fs = scanText('const q = "SELECT * FROM users WHERE id = " + userId;');
  assert.ok(has(fs, "SQL001"));
});

test("template literal interpolation flagged as SQL002", () => {
  const fs = scanText("const q = `SELECT * FROM t WHERE name = '${name}'`;");
  assert.ok(has(fs, "SQL002"));
});

test("printf-style formatting flagged as SQL003", () => {
  const fs = scanText('q = "SELECT * FROM t WHERE id = %s" % value');
  assert.ok(has(fs, "SQL003"));
});

test("non-SQL concatenation ignored", () => {
  const fs = scanText('const greeting = "hello " + name;');
  assert.equal(fs.length, 0);
});

test("parameterized query is clean", () => {
  const fs = scanText('db.query("SELECT * FROM users WHERE id = $1", [userId]);');
  assert.equal(fs.length, 0);
});

test("line numbers reported", () => {
  const fs = scanText('const x = 1;\n\nconst q = "SELECT a FROM t WHERE b=" + v;');
  assert.equal(fs[0].line, 3);
});

test("empty input yields no findings", () => {
  assert.equal(scanText("").length, 0);
});
